File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    bool isEnough(vector<int>& piles, int h, int s) {
        long long time = 0;
        for (int i : piles) {
            if (i % s == 0) {
                time += i / s;
            } else {
                time += (i / s + 1);
            }
        }
        return h >= time;
    }
    int minEatingSpeed(vector<int>& piles, int h) {
        if (piles.size() == 1) {
            return (piles[0] + h - 1) / h;
        }


        sort(piles.begin(), piles.end());
        int left = 1;
        int right = piles[piles.size() - 1];
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (isEnough(piles, h, mid)) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # def canEat(k) -> bool:
        #     return sum((b + k - 1) // k for b in piles) <= h

        # minSpeed = 1
        # maxSpeed = max(piles)

        # while minSpeed <= maxSpeed:
        #     midSpeed = (minSpeed + maxSpeed) // 2

        #     if canEat(midSpeed):
        #         maxSpeed = midSpeed - 1
        #     else:
        #         minSpeed = midSpeed + 1

        # return minSpeed

        minSpeed, maxSpeed = max(sum(piles) // h, 1), max(piles)

        def canEat(speed):
            return sum((pile + (speed - 1)) // speed for pile in piles) <= h

        while minSpeed <= maxSpeed:
            midSpeed = (minSpeed + maxSpeed) // 2

            if canEat(midSpeed):
                maxSpeed = midSpeed - 1
            else:
                minSpeed = midSpeed + 1

        return minSpeed